File: examples/binary_fusion_tap.py
#!/usr/bin/env python3
"""
Binary Tap Operation - Python Implementation
Deterministic key generation using bit-shifting and XOR operations

This demonstrates basic binary manipulation operations:
- Bit-shifting (multiplication by powers of 2)
- Addition (introducing offset bits)
- XOR (extracting difference bits)
"""

def binary_fusion_tap(k: int) -> dict:
    """
    Generate binary tap operation with bit-shifting and XOR extraction.

    This performs a simple mathematical transformation:
    1. Create seed from digit sequence (1,2,3,...,k)
    2. Shift bits left by 3 positions (multiply by 8)
    3. Add offset parameter k
    4. Use XOR to extract the bits contributed by k

    Args:
        k: Offset parameter (recommended: 11 for this application)

    Returns:
        Dictionary with transformation data
    """
    # 1. Generate seed from concatenated digit sequence
    seed_val = int("".join(map(str, range(1, k + 1))))

    # 2. Apply bit-shift left by 3 (equivalent to multiplying by 8)
    shifted_val = seed_val << 3

    # 3. Add offset to shifted value
    result = shifted_val + k

    # 4. Extract difference bits using XOR
    # For k < 10, no extraction is needed
    # For k >= 10, XOR reveals the bits added by the offset
    if k < 10:
        diff_bits = 0
    else:
        diff_bits = result ^ (seed_val * 8)

    return {
        "k": k,
        "seed_value": seed_val,
        "binary_seed": bin(seed_val),
        "tap_state": bin(result),
        "zpe_overflow": bin(diff_bits),  # Kept for backward compatibility
        "zpe_overflow_decimal": diff_bits  # Kept for backward compatibility
    }


# Example usage
if __name__ == "__main__":
    result = binary_fusion_tap(11)
    print(f"Offset Parameter: {result['k']}")
    print(f"Seed Value: {result['seed_value']}")
    print(f"Result State: {result['tap_state']}")
    print(f"Difference Bits: {result['zpe_overflow']}")

File: examples/binary_fusion_tap.c
/**
 * Binary Fusion Tap - C Implementation
 * Quantum-inspired key generation using 8-fold Heartbeat and ZPE Overflow
 *
 * Note: This implementation uses uint64_t. For larger k values,
 * consider using a big integer library like GMP.
 */

#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>

typedef struct {
    int k;
    uint64_t seed_value;
    uint64_t tap_state;
    uint64_t zpe_overflow;
} BinaryFusionResult;

/**
 * Generate seed value from concatenated sequence
 * For k <= 15 to avoid overflow with uint64_t
 */
uint64_t generate_seed(int k) {
    char seed_str[256] = {0};
    char temp[16];

    for (int i = 1; i <= k; i++) {
        sprintf(temp, "%d", i);
        strcat(seed_str, temp);
    }

    return strtoull(seed_str, NULL, 10);
}

/**
 * Binary Fusion Tap algorithm
 * k: Tap parameter (recommended: 11, max: 15 for uint64_t)
 */
BinaryFusionResult binary_fusion_tap(int k) {
    BinaryFusionResult result;
    result.k = k;

    // 1. Generate seed from concatenated sequence
    result.seed_value = generate_seed(k);

    // 2. Apply 8-fold Heartbeat (bit-shift left by 3)
    uint64_t heartbeat_val = result.seed_value << 3;

    // 3. Add Phase Offset
    result.tap_state = heartbeat_val + k;

    // 4. Extract ZPE Overflow
    if (k < 10) {
        result.zpe_overflow = 0;
    } else {
        result.zpe_overflow = result.tap_state ^ (result.seed_value * 8);
    }

    return result;
}

/**
 * Print result in binary format
 */
void print_binary(uint64_t n) {
    if (n > 1) {
        print_binary(n / 2);
    }
    printf("%llu", n % 2);
}

int main() {
    BinaryFusionResult result = binary_fusion_tap(11);

    printf("K Parameter: %d\n", result.k);
    printf("Seed Value: %llu\n", result.seed_value);
    printf("Tap State: 0b");
    print_binary(result.tap_state);
    printf("\n");
    printf("ZPE Overflow: 0b");
    print_binary(result.zpe_overflow);
    printf(" (decimal: %llu)\n", result.zpe_overflow);

    return 0;
}
